File: BTPostIn.java
// Time Complexity : O(n)
// Space Complexity : O(maxDepth)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

import java.util.Arrays;
import java.util.HashMap;

public class BTPostIn {
    //With extra Space
    public TreeNode buildTreeExtraSpace(int[] inorder, int[] postorder) {
        if (postorder.length == 0) {
            return null;
        }

        TreeNode root = new TreeNode(postorder[postorder.length - 1]);

        int index = -1;
        for (int i = 0; i < inorder.length; i++) {
            if (inorder[i] == root.val) {
                index = i;
            }
        }

        int[] inorder_left = Arrays.copyOfRange(inorder, 0, index);
        int[] postorder_left = Arrays.copyOfRange(postorder, 0, index);
        int[] inorder_right = Arrays.copyOfRange(inorder, index + 1, inorder.length);
        int[] postorder_right = Arrays.copyOfRange(postorder, index, postorder.length - 1);

        root.left = buildTree(inorder_left, postorder_left);
        root.right = buildTree(inorder_right, postorder_right);

        return root;
    }

    // Without copying arrays and adding hashmap
    HashMap<Integer, Integer> map = new HashMap<>();

    int currRoot;

    public TreeNode buildTree(int[] inorder, int[] postorder) {

        currRoot = postorder.length - 1;

        for (int i = 0; i < inorder.length; i++) {
            // No duplicates, hence just put
            map.put(inorder[i], i);
        }

        return recurrence(inorder, postorder, 0, inorder.length - 1);

    }

    private TreeNode recurrence(int[] inorder, int[] postorder, int start, int end) {

        // base
        if (start > end || currRoot < 0) {
            return null;
        }

        // recursive case
        TreeNode root = new TreeNode(postorder[currRoot]);
        currRoot -= 1;

        int index = map.get(root.val);

        root.right = recurrence(inorder, postorder, index + 1, end);
        root.left = recurrence(inorder, postorder, start, index - 1);

        return root;

    }
}

File: bt_post_in.py
#Time Complexity: O(N).
#Auxiliary Space: O(N)
#Did this code successfully run on Leetcode :Yes


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.pre_idx = 0
        self.inorder_map = {}
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        def array_tree(left,right):
            if left>=right:
                return
            root_value = postorder[self.pre_idx]
            root_index = self.inorder_map[root_value]
            root = TreeNode(root_value)
            self.pre_idx -=1
            
            
            root.right = array_tree(root_index+1,right)
            root.left = array_tree(left,root_index)
            return root
                

        self.pre_idx = len(postorder)-1
        for index,element in enumerate(inorder):
            self.inorder_map[element] = index
            
        return array_tree(0,len(postorder))